src/test/ui/async-await/issue-70818.rs

```diff
@@ -0,0 +1,9 @@
+// edition:2018
```

Comment:
Can you add a test that includes the same type being captured as an upvar _and_ held across an await in the interior?

Alternatively, we can refactor the code to only handle one case or the other, which would simplify things a bit. (We could use an enum instead of two `Option`s, for instance.)